syntax = "proto3";

package statistics;

service StatisticsService {
    rpc GetPostStats(PostRequest) returns (PostStatsResponse);
    rpc GetViewsDynamic(PostRequest) returns (DynamicResponse);
    rpc GetLikesDynamic(PostRequest) returns (DynamicResponse);
    rpc GetCommentsDynamic(PostRequest) returns (DynamicResponse);
    rpc GetTopPosts(TopRequest) returns (TopPostsResponse);
    rpc GetTopUsers(TopRequest) returns (TopUsersResponse);
}

message PostRequest {
    string post_id = 1;
}


message PostStatsResponse {
    uint64 views = 1;
    uint64 likes = 2;
    uint64 comments = 3;
}

message DynamicData {
    string date = 1;
    uint64 count = 2;
}

message DynamicResponse {
    repeated DynamicData data = 1;
}

enum Metric {
    VIEWS = 0;
    LIKES = 1;
    COMMENTS = 2;
}

message TopRequest {
    Metric metric = 1;
}

message TopPost {
    string post_id = 1;
    uint64 count = 2;
}

message TopPostsResponse {
    repeated TopPost posts = 1;
}

message TopUser {
    string user_id = 1;
    uint64 count = 2;
}

message TopUsersResponse {
    repeated TopUser users = 1;
}